fix(make-dedicated-lockfile): error with workspace protocol

<!--
  If this issue affects many people in a company/big team, create a post for your company in the following discussion:
  https://github.com/pnpm/pnpm/discussions/3787
  and link the issue in your post.

  This will help us prioritize issues that affect more people.
-->

### pnpm version: 7.5.1

### Code to reproduce the issue:

```sh
$ git clone git@github.com:jjangga0214/haetae.git
$ cd haetae 
$ git reset --hard 90c6718
$ pnpm install
$ pnpm add -wD @pnpm/make-dedicated-lockfile
$ cd packages/cli
$ pnpm make-dedicated-lockfile
```

### Expected behavior:

It creates `pnpm-lock.yaml` in packages/core.

### Actual behavior:

An error happens.

Note that `@haetae/cli`(packages/cli) depends on `@haetae/core`(packages/core) by workspace protocol.

```sh
➜  cli git:(main) ✗ pnpm make-dedicated-lockfile              
(node:66811) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./src/" in the "exports" field module resolution of the package at /Users/ocean/main/haetae/node_modules/.pnpm/@pnpm+exportable-manifest@3.0.7/node_modules/ramda/package.json.
Update this package.json to use a subpath pattern like "./src/*".
(Use `node --trace-deprecation ...` to show where the warning was created)
(node:66811) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./src/" in the "exports" field module resolution of the package at /Users/ocean/main/haetae/node_modules/.pnpm/@pnpm+lockfile-file@5.1.4_@pnpm+logger@4.0.0/node_modules/ramda/package.json.
Update this package.json to use a subpath pattern like "./src/*".
(node:66811) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./src/" in the "exports" field module resolution of the package at /Users/ocean/main/haetae/node_modules/.pnpm/@pnpm+prune-lockfile@4.0.11/node_modules/ramda/package.json.
Update this package.json to use a subpath pattern like "./src/*".
(node:66811) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./src/" in the "exports" field module resolution of the package at /Users/ocean/main/haetae/node_modules/.pnpm/@pnpm+make-dedicated-lockfile@0.3.14/node_modules/ramda/package.json.
Update this package.json to use a subpath pattern like "./src/*".
/Users/ocean/main/haetae/node_modules/.pnpm/@pnpm+exportable-manifest@3.0.7/node_modules/@pnpm/exportable-manifest/lib/index.js:57
            throw new error_1.default('CANNOT_RESOLVE_WORKSPACE_PROTOCOL', `Cannot resolve workspace protocol of dependency "${depName}" ` +
                  ^

PnpmError: Cannot resolve workspace protocol of dependency "@haetae/core" because this dependency is not installed. Try running "pnpm install".
    at makePublishDependency (/Users/ocean/main/haetae/node_modules/.pnpm/@pnpm+exportable-manifest@3.0.7/node_modules/@pnpm/exportable-manifest/lib/index.js:57:19)
    at async /Users/ocean/main/haetae/node_modules/.pnpm/@pnpm+exportable-manifest@3.0.7/node_modules/@pnpm/exportable-manifest/lib/index.js:44:9
    at async Promise.all (index 0)
    at async makePublishDependencies (/Users/ocean/main/haetae/node_modules/.pnpm/@pnpm+exportable-manifest@3.0.7/node_modules/@pnpm/exportable-manifest/lib/index.js:41:61)
    at async makePublishManifest (/Users/ocean/main/haetae/node_modules/.pnpm/@pnpm+exportable-manifest@3.0.7/node_modules/@pnpm/exportable-manifest/lib/index.js:26:22)
    at async default_1 (/Users/ocean/main/haetae/node_modules/.pnpm/@pnpm+make-dedicated-lockfile@0.3.14/node_modules/@pnpm/make-dedicated-lockfile/lib/index.js:49:29)
    at async main (/Users/ocean/main/haetae/node_modules/.pnpm/@pnpm+make-dedicated-lockfile@0.3.14/node_modules/@pnpm/make-dedicated-lockfile/lib/bin.js:16:5) {
  code: 'ERR_PNPM_CANNOT_RESOLVE_WORKSPACE_PROTOCOL',
  hint: undefined,
  attempts: undefined
}
```

Thank you so much for pnpm.

### Additional information:

 - `node -v` prints: v16.13.1
 - Windows, macOS, or Linux?: macOS

--- a/.changeset/empty-apricots-tie.md
+++ b/.changeset/empty-apricots-tie.md
@@ -1,0 +1,5 @@
+---
+"@pnpm/read-project-manifest": patch
+---
+
+It should be possible to rewrite a manifest back to its initial content [#5061](https://github.com/pnpm/pnpm/issues/5061).

--- a/.changeset/famous-humans-hunt.md
+++ b/.changeset/famous-humans-hunt.md
@@ -1,0 +1,5 @@
+---
+"@pnpm/make-dedicated-lockfile": patch
+---
+
+Read the dependency versions before moving node_modules.

--- a/packages/make-dedicated-lockfile/src/index.ts
+++ b/packages/make-dedicated-lockfile/src/index.ts
@@ -34,6 +34,11 @@ export default async function (lockfileDir: string, projectDir: string) {
   const dedicatedLockfile = pruneSharedLockfile(lockfile)
 
   await writeWantedLockfile(projectDir, dedicatedLockfile)
+
+  const { manifest, writeProjectManifest } = await readProjectManifest(projectDir)
+  const publishManifest = await exportableManifest(projectDir, manifest)
+  await writeProjectManifest(publishManifest)
+
   const modulesDir = path.join(projectDir, 'node_modules')
   const tmp = path.join(projectDir, 'tmp_node_modules')
   const tempModulesDir = path.join(projectDir, 'node_modules/.tmp')
@@ -45,10 +50,6 @@ export default async function (lockfileDir: string, projectDir: string) {
   } catch (err: any) { // eslint-disable-line
     if (err['code'] !== 'ENOENT') throw err
   }
-
-  const { manifest, writeProjectManifest } = await readProjectManifest(projectDir)
-  const publishManifest = await exportableManifest(projectDir, manifest)
-  await writeProjectManifest(publishManifest)
 
   try {
     await pnpmExec([

--- a/packages/make-dedicated-lockfile/test/fixtures/fixture/packages/is-negative/package.json
+++ b/packages/make-dedicated-lockfile/test/fixtures/fixture/packages/is-negative/package.json
@@ -4,6 +4,6 @@
   "dependencies": {
     "ramda": "0.26.0",
     "request": "^2.0.0",
-    "is-positive": "workspace:1.0.0"
+    "is-positive": "workspace:^"
   }
 }

--- a/packages/make-dedicated-lockfile/test/index.ts
+++ b/packages/make-dedicated-lockfile/test/index.ts
@@ -10,7 +10,7 @@ const f = fixtures(__dirname)
 test('makeDedicatedLockfile()', async () => {
   const tmp = f.prepare('fixture')
   fs.writeFileSync('.npmrc', 'store-dir=store\ncache-dir=cache', 'utf8')
-  await pnpmExec(['install'], { cwd: tmp })
+  await pnpmExec(['install', '--no-frozen-lockfile'], { cwd: tmp })
   const projectDir = path.join(tmp, 'packages/is-negative')
   await makeDedicatedLockfile(tmp, projectDir)
 

--- a/packages/make-dedicated-lockfile/test/index.ts
+++ b/packages/make-dedicated-lockfile/test/index.ts
@@ -1,5 +1,6 @@
 import fs from 'fs'
 import path from 'path'
+import pnpmExec from '@pnpm/exec'
 import { readWantedLockfile } from '@pnpm/lockfile-file'
 import fixtures from '@pnpm/test-fixtures'
 import makeDedicatedLockfile from '../lib'
@@ -9,6 +10,7 @@ const f = fixtures(__dirname)
 test('makeDedicatedLockfile()', async () => {
   const tmp = f.prepare('fixture')
   fs.writeFileSync('.npmrc', 'store-dir=store\ncache-dir=cache', 'utf8')
+  await pnpmExec(['install'], { cwd: tmp })
   const projectDir = path.join(tmp, 'packages/is-negative')
   await makeDedicatedLockfile(tmp, projectDir)
 

--- a/packages/read-project-manifest/src/index.ts
+++ b/packages/read-project-manifest/src/index.ts
@@ -179,14 +179,16 @@ function createManifestWriter (
     manifestPath: string
   }
 ): (WriteProjectManifest) {
-  const initialManifest = normalize(JSON.parse(JSON.stringify(opts.initialManifest)))
+  let initialManifest = normalize(opts.initialManifest)
   return async (updatedManifest: ProjectManifest, force?: boolean) => {
     updatedManifest = normalize(updatedManifest)
     if (force === true || !equal(initialManifest, updatedManifest)) {
-      return writeProjectManifest(opts.manifestPath, updatedManifest, {
+      await writeProjectManifest(opts.manifestPath, updatedManifest, {
         indent: opts.indent,
         insertFinalNewline: opts.insertFinalNewline,
       })
+      initialManifest = normalize(updatedManifest)
+      return Promise.resolve(undefined)
     }
     return Promise.resolve(undefined)
   }
@@ -200,6 +202,7 @@ const dependencyKeys = new Set([
 ])
 
 function normalize (manifest: ProjectManifest) {
+  manifest = JSON.parse(JSON.stringify(manifest))
   const result = {}
 
   for (const key of Object.keys(manifest)) {

--- a/packages/read-project-manifest/test/index.ts
+++ b/packages/read-project-manifest/test/index.ts
@@ -173,3 +173,17 @@ test('preserve trailing new line at the end of package.json5', async () => {
   const rawManifest = await fs.readFile('package.json5', 'utf8')
   expect(rawManifest).toBe("{dependencies:{bar:'1.0.0'}}")
 })
+
+test('canceling changes to a manifest', async () => {
+  process.chdir(tempy.directory())
+
+  await fs.writeFile('package.json', JSON.stringify({ name: 'foo' }), 'utf8')
+
+  const { writeProjectManifest } = await readProjectManifest(process.cwd())
+
+  await writeProjectManifest({ name: 'bar' })
+  expect(await fs.readFile('package.json', 'utf8')).toBe(JSON.stringify({ name: 'bar' }))
+
+  await writeProjectManifest({ name: 'foo' })
+  expect(await fs.readFile('package.json', 'utf8')).toBe(JSON.stringify({ name: 'foo' }))
+})

